File: Problems/0839. (H) Similar String Groups/Python3/solution.py
class Solution:
  def numSimilarGroups(self, strs: List[str]) -> int:
    # iterative DFS

    # generate graph
    n = len(strs)
    m = len(strs[0])
    adj = [[] for _ in range(n)]
    for i in range(n):
      for j in range(i, n):
        diff = 0
        for k in range(m):
          if strs[i][k] != strs[j][k]:
            diff += 1
        if not diff or diff == 2:
          adj[i].append(j)
          adj[j].append(i)
    
    # count connected components
    visited = [False] * n
    ret = 0
    for i in range(n):
      if visited[i]:
        continue
      ret += 1
      nodes = []
      nodes.append(i)
      while nodes:
        cur = nodes.pop()
        if visited[cur]:
          continue
        visited[cur] = True
        for nxt in adj[cur]:
          if not visited[nxt]:
            nodes.append(nxt)
      
    return ret

File: Problems/0839. (H) Similar String Groups/Java/solution.java
class Solution {
  public int numSimilarGroups(String[] strs) {
    // iterative DFS

    // generate graph
    int n = strs.length;
    int m = strs[0].length();
    List<Integer>[] adj = new ArrayList[n];
    for (int i = 0; i < n; i++)
      adj[i] = new ArrayList();
    for (int i = 0; i < n; i++) {
      for (int j = i; j < n; j++) {
        int diff = 0;
        for (int k = 0; k < m; k++) {
          if (strs[i].charAt(k) != strs[j].charAt(k))
            diff++;
        }
        if (diff == 0 || diff == 2) {
          adj[i].add(j);
          adj[j].add(i);
        }
      }
    }

    // count connected components
    boolean[] visited = new boolean[n];
    int ret = 0;
    Deque<Integer> nodes = new ArrayDeque();
    for (int i = 0; i < n; i++) {
      if (visited[i])
        continue;
      ret++;
      nodes.push(i);
      while (nodes.size() > 0) {
        int cur = nodes.pop();
        if (visited[cur])
          continue;
        visited[cur] = true;
        for (int nxt: adj[cur]) {
          if (!visited[nxt])
            nodes.push(nxt);
        }
      }
    }
    return ret;
  }
}
